frontend/junior-2/react.md

```diff
@@ -1,63 +1,58 @@
+            Click me
+          </button>
+        </div>
+      );
+    }
+    ```
+  * Как сделать так, что бы все 5 раз вывелось последнее значение, то есть цифра 5?
+  * Что выведется в консоли для следующего кода, если несколько раз быстро нажать на кнопку?
+    ```javascript
+    function Example() {
+      const [count, setCount] = useState(0);
+
+      useEffect(() => {
+        console.log(count);
+        if (count % 2 === 0) {
+          return () => setTimeout(() => console.log("It was even render"), 1000);
+        } else {
+          return () => setTimeout(() => console.log("It was odd render"), 1000);
+        }
+      });
+
+      return (
+        <div>
+          <p>You clicked {count} times</p>
+          <button onClick={() => setCount(count + 1)}>Click me</button>
+        </div>
+      );
+    }
     ```
-* Что такое Virtual DOM и зачем он нужен?
-* Что такое `state` и `props` у компонента?
-  * Как обновить `state`? Почему не стоит это делать через прямую запись в `this.state`?
-  * Если передать в `setState` не все свойства, которые до этого уже были в стейте, то будут ли удалены те, что не переданы?
-  * В каких случаях стоит хранить значения в стейте компоненты, а в каких случаях в инстансе?
-  * Как правильно обновлять `state` на основе предыдущего значения `state`? Какие могут при этом возникнуть проблемы?
-  * Что такое top-down data flow? Что является основным механизмом синхронизации нескольких компонент?
-* `props.children`
-  * Что это такое?
-  * Когда может быть полезным?
-  * Что позволяет делать `React.Children` API?
-* Component Lifecycle
-  * Что это такое?
-  * Когда какие методы вызываются? В каком порядке?
-  * Как и в каких случаях использовать те или иные методы жизненного цикла?
-* Что должен делать метод `render`? Почему необходимо держать этот метод чистым?
-* Hooks
-  * Какие существуют правила при работе с хуками?
-  * Для чего предназначены хуки `useState`, `useReducer`, `useEffect`? Как с ними работать?
-  * Что может возвращать колбэк, передаваемый в `useEffect`?
-  * Что такое массив зависимостей? Рассказать как он влияет на работу хука на примере useEffect (в том числе на возвращаемое колбэком значение).
-* Что такое HOC? Когда хоки могут быть полезны?
-* Что такое React Fragment? Для чего он нужен? Как его использовать?
-* Performance
-  * Когда передача инлайн-коллбека ухудшает производительность и почему? Пример: `<LoginButton onClick={(e) => this.handleClick(e, user)}>`
-* refs
-  * Что это такое? Зачем нужны? Как их использовать?
-  * Как и для чего использовать хук `useRef`?
-  * Что делает и зачем нужен метод `React.forwardRef`?
-  * Почему refs стоит использовать умеренно?
-* Controlled и uncontrolled компоненты
-  * Что это такое? Как происходит работа с инпутами?
-  * Как задать дефолтное значение для инпута в uncontrolled компоненте?
-* Что такое Context и когда может быть полезен?
+* Как и для чего использовать хуки `useMemo` и `useCallback`? 
```

Comment:
Предлагаю добавить еще несколько вопросов по хукам, есть ощущение, что тема хуков в карте не раскрыта )
Например, вопросы, касающиеся производительности:

- Являются ли хуки менее предпочтительными с точки зрения производительности (из-за того, что на каждый рендер создается новая функция)?
- Всегда ли следует оборачивать код в `useCallback`/`useMemo`?

Также можно добавить вопрос про разницу между `useEffect` и `componentDidMount`, правда пока непонятно, куда его лучше впихнуть.

Это то, что сходу вспомнилось, можно еще подумать, если решим, что блок с хуками нужно расширять. Что думаете?

Reply:
Я не против расширить, но пока не вижу чем. Как буд-то бы сильно не разгуляешься. Например массив зависимостей во всех хуках работает одинаково, спрашивать про него ещё раз нет смысла. 
**Upd:** Впрочем...ведь не одними `useEffect`, `useCallback` и `useMemo` богаты..может можно добавить вопросы про другие базовые хуки. Да да да... я гляну.

> Являются ли хуки менее предпочтительными с точки зрения производительности (из-за того, что на каждый рендер создается новая функция)?

Если честно немного не понял вопроса. Это в контексте сравнения с классовыми компонентами? Но так-то все уже перешли на функциональные компоненты. А если это вопрос про функциональные компоненты..разве есть альтернативы хукам?

> Всегда ли следует оборачивать код в useCallback/useMemo?

Тут согласен, можно добавить.

> Также можно добавить вопрос про разницу между `useEffect` и `componentDidMount`

Яяяя...не уверен) Просто когда люди переходят на функциональные компоненты, во всех гайдах кричат о том, мол не думайте о `useEffect` как о методах жизненного цикла. Что сравнивать их так себе идея. Надо подумать.

Reply:
> Если честно немного не понял вопроса. Это в контексте сравнения с классовыми компонентами? Но так-то все уже перешли на функциональные компоненты. А если это вопрос про функциональные компоненты..разве есть альтернативы хукам?

На самом деле, это вопрос [прям из доки](https://ru.reactjs.org/docs/hooks-faq.html#are-hooks-slow-because-of-creating-functions-in-render), я просто его перефразировала, видимо не слишком удачно )

Reply:
> Тут согласен, можно добавить.

Есть пара неплохих статей на эту тему на Хабре, можно будет тогда в качестве источников добавить.

Reply:
> Яяяя...не уверен) Просто когда люди переходят на функциональные компоненты, во всех гайдах кричат о том, мол не думайте о useEffect как о методах жизненного цикла. Что сравнивать их так себе идея. Надо подумать.

Ок, я возможно отстала от жизни )

Reply:
На основе нашего обсуждения сделал коммит ниже)

Reply:
> Есть пара неплохих статей на эту тему на Хабре, можно будет тогда в качестве источников добавить.
> 
> UPD. [Вот](https://habr.com/ru/post/529950/) и [вот](https://habr.com/ru/company/domclick/blog/656781/) )

Я бы еще отдельной ссылкой добавила этот [коммент](https://habr.com/ru/post/529950/#comment_22380330) из первой статьи

Reply:
> Я бы еще отдельной ссылкой добавила этот [коммент](https://habr.com/ru/post/529950/#comment_22380330) из первой статьи

Done)